lib/active_merchant/billing/gateways/adyen.rb

```diff
@@ -124,14 +127,30 @@ def add_shopper_interaction(post, payment, options={})
       def add_address(post, options)
         return unless post[:card] && post[:card].kind_of?(Hash)
         if (address = options[:billing_address] || options[:address]) && address[:country]
-          post[:card][:billingAddress] = {}
-          post[:card][:billingAddress][:street] = address[:address1] || 'N/A'
-          post[:card][:billingAddress][:houseNumberOrName] = address[:address2] || 'N/A'
-          post[:card][:billingAddress][:postalCode] = address[:zip] if address[:zip]
-          post[:card][:billingAddress][:city] = address[:city] || 'N/A'
-          post[:card][:billingAddress][:stateOrProvince] = address[:state] if address[:state]
-          post[:card][:billingAddress][:country] = address[:country] if address[:country]
+          post[:billingAddress] = {}
+          post[:billingAddress][:street] = address[:address1] || 'N/A'
+          post[:billingAddress][:houseNumberOrName] = address[:address2] || 'N/A'
+          post[:billingAddress][:postalCode] = address[:zip] if address[:zip]
+          post[:billingAddress][:city] = address[:city] || 'N/A'
+          post[:billingAddress][:stateOrProvince] = address[:state] if address[:state]
+          post[:billingAddress][:country] = address[:country] if address[:country]
         end
+        if shipping_address = options[:shipping_address] || address
```

Comment:
_Nothing to do with code logic, just code cleanup:_

With these two conditionals, your method is growing and the name isn't representative of what it's doing. Have you thought of splitting into two disctinct methods? `add_billing_address` and `add_shipping_address`?